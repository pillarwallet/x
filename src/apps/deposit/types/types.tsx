export type BalanceInfo = {
  type?: 'BalanceInfo';
  name: string;
  chain: string;
  address: string;
  decimals: number;
  symbol: string;
  logoURI: string;
  balance: string;
};

export type AddedAssets = {
  type?: 'AddedAsset';
  assetType: 'token' | 'nft';
  tokenId?: string;
  chain: string;
  tokenAddress: string;
  balance: number;
};

export type Network =
  | 'ethereum'
  | 'polygon'
  | 'gnosis'
  | 'base'
  | 'bnb smart chain'
  | 'optimism'
  | 'arbitrum';

export type TokenList = {
  address: string;
  chainId: number;
  decimals: string;
  name: string;
  symbol: string;
  logoURI: string;
};

export type BalancesByChain = {
  [chain: string]: BalanceInfo[];
};
