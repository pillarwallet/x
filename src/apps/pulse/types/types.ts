export type TransactionStatusState =
  | 'Starting Transaction'
  | 'Transaction Pending'
  | 'Transaction Complete'
  | 'Transaction Failed';

export type TransactionStep =
  | 'Submitted'
  | 'Pending'
  | 'ResourceLock'
  | 'Completed';

export type StepStatus = 'completed' | 'pending' | 'failed' | 'inactive';

export interface TokenInfo {
  symbol: string;
  name: string;
  logo: string;
  address?: string;
}

export interface SellOffer {
  tokenAmountToReceive: number;
  minimumReceive: number;
}

export interface PayingToken {
  totalUsd: number;
  name: string;
  symbol: string;
  logo: string;
  actualBal: string;
  totalRaw: string;
  chainId: number;
  address: string;
}

export interface TokenDetails {
  symbol: string;
  name: string;
  address: string;
  chainId: number;
  amount: string;
  logo: string;
  type: 'BUY_TOKEN' | 'SELL_TOKEN';
}

export interface BuyModeDetails {
  usdAmount: string;
  payingTokens: PayingToken[];
  totalPayingUsd: number;
}

export interface TransactionStatusConfig {
  icon: string;
  containerClasses: string;
  iconClasses: string;
  color: string;
}

export interface ButtonConfig {
  bgColor: string;
  textColor: string;
  borderColor: string;
  label: string;
}

export interface ProgressStepConfig {
  label: string;
  order: number;
}

export interface TransactionStatusProps {
  closeTransactionStatus: () => void;
  userOpHash: string;
  chainId: number;
  gasFee?: string;
  // Transaction data from PreviewSell/PreviewBuy
  isBuy?: boolean;
  sellToken?: TokenInfo | null;
  buyToken?: TokenInfo | null;
  tokenAmount?: string;
  sellOffer?: SellOffer | null;
  payingTokens?: PayingToken[];
  usdAmount?: string;
  // Externalized polling state
  currentStatus: TransactionStatusState;
  errorDetails: string;
  submittedAt?: Date;
  pendingCompletedAt?: Date;
  blockchainTxHash?: string;
  resourceLockTxHash?: string;
  completedTxHash?: string;
  completedChainId?: number;
  resourceLockChainId?: number;
  resourceLockCompletedAt?: Date;
  isResourceLockFailed?: boolean;
}

export interface TransactionDetailsProps {
  onDone: () => void;
  userOpHash: string;
  chainId: number;
  status: TransactionStatusState;
  // Transaction data from PreviewSell/PreviewBuy
  isBuy?: boolean;
  sellToken?: TokenInfo | null;
  buyToken?: TokenInfo | null;
  tokenAmount?: string;
  sellOffer?: SellOffer | null;
  payingTokens?: PayingToken[];
  usdAmount?: string;
  submittedAt?: Date;
  pendingCompletedAt?: Date;
  resourceLockCompletedAt?: Date;
  txHash?: string;
  gasFee?: string;
  errorDetails?: string;
  resourceLockTxHash?: string;
  completedTxHash?: string;
  resourceLockChainId?: number;
  completedChainId?: number;
  isResourceLockFailed?: boolean;
}

export interface TransactionInfoProps {
  status: TransactionStatusState;
  userOpHash: string;
  txHash?: string;
  chainId: number;
  gasFee?: string;
  completedAt?: Date;
  // Buy-specific fields
  isBuy?: boolean;
  resourceLockTxHash?: string;
  resourceLockChainId?: number;
  completedTxHash?: string;
  completedChainId?: number;
}

export interface ProgressStepProps {
  step: TransactionStep;
  status: StepStatus;
  label: string;
  isLast?: boolean;
  showLine?: boolean;
  lineStatus?: StepStatus;
  timestamp?: string | React.ReactNode;
}

export interface TransactionErrorBoxProps {
  technicalDetails?: string;
}

export interface UseClickOutsideOptions {
  ref: React.RefObject<HTMLElement>;
  callback: () => void;
  condition: boolean;
}

export interface UseKeyboardNavigationOptions {
  onEscape: () => void;
  onEnter?: () => void;
  enabled?: boolean;
}

export interface TechnicalDetails {
  transactionType: 'BUY' | 'SELL';
  transactionHash: string;
  hashType: 'bidHash' | 'userOpHash';
  chainId: number;
  status: TransactionStatusState;
  timestamp: string;
  accountAddress: string;
  token: TokenDetails | null;
  sellOffer: SellOffer | null;
  buyMode: BuyModeDetails | null;
  transactionHashes: {
    [key: string]: string;
  };
  chains: {
    mainChainId: number;
    resourceLockChainId: string | number;
    completedChainId: string | number;
  };
  timestamps: {
    [key: string]: string;
  };
  error: {
    details: string;
    isResourceLockFailed: boolean;
    failureStep: string;
  };
  gas: {
    fee: string;
  };
  stepStatus: {
    [key: string]: StepStatus;
  };
}
