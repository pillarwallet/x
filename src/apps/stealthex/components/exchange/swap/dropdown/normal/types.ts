
import type React from 'react';
import { StructuredCurrency } from '../../../../../type';
import { CurrencyData } from '../../../../../lib/backend/api';

export type ClickCallback = (currency: CurrencyData) => void;

export type AvailableCurrency =
  | {
    availableCurrency?: Omit<StructuredCurrency, 'fiat'> & {
      all?: CurrencyData[];
    };
    type: 'get';
  }
  | {
    availableCurrency?: StructuredCurrency & {
      all?: CurrencyData[];
    };
    type: 'send';
  };

export type NormalDropdownProps = AvailableCurrency & {
  loading?: boolean;
  selectedCurrency: CurrencyData | null;
  searchContext: string;
  fiat?: boolean;
  onClick?: ClickCallback;
  components?: {
    NotFound?: React.FC<React.PropsWithChildren>;
    Group?: React.FC<React.PropsWithChildren>;
    Item?: React.FC<
      React.PropsWithChildren<
        | { loading: true }
        | {
          loading?: false;
          currency: CurrencyData;
          selected?: boolean;
          onClick?: () => void;
        }
      >
    >;
  };
  groupHeight?: number;
  itemHeight?: number;
  maxHeight?: number;
  embed?: boolean;
};
