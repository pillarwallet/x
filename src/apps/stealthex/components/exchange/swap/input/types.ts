
import { CurrencyData } from '../../../../lib/backend/api';
import type { AvailableCurrency } from '../dropdown/normal/types';

export type SavedAvailableCurrency =
  | {
      floatingDropdown: true;
      availableCurrency: CurrencyData[];
      type: 'get' | 'send';
    }
  | ({
      floatingDropdown?: false;
    } & AvailableCurrency);

export type InputProps = {
  currency: CurrencyData;
  value?: string;
  loading?: boolean;
  loadingCurrency?: boolean;
  error?: string | JSX.Element;
  fiat?: boolean;
  disabled?: boolean;
  withLock?: boolean | null;
  widget?: boolean;
  fixed?: boolean;
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
  onFixedChange?: (fixed: boolean) => void;
  onCurrencySelect?: (currency: CurrencyData) => void;
  pressable?: boolean;
} & SavedAvailableCurrency;
