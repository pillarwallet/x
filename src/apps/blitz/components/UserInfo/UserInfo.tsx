import {
  Copy as CopyIcon,
  CopySuccess as CopySuccessIcon,
} from 'iconsax-react';
import { useCallback, useEffect, useState } from 'react';
import CopyToClipboard from 'react-copy-to-clipboard';

// utils

// components
import RandomAvatar from '../RandomAvatar/RandomAvatar';
import Body from '../Typography/Body';

const UserInfo = () => {
  const DUMMY_WALLET_ADDRESS = '0x4a5e60B1F60E1A23B45C67D89E0F123456789ABCD';
  const DUMMY_USERNAME = 'Tester124';

  const [copied, setCopied] = useState<boolean>(false);

  const onCopyCodeClick = useCallback(() => {
    if (copied) {
      setCopied(false);
      return;
    }

    setCopied(true);
  }, [copied]);

  useEffect(() => {
    const codeCopyActionTimeout = setTimeout(() => {
      setCopied(false);
    }, 500);

    return () => {
      clearTimeout(codeCopyActionTimeout);
    };
  }, [copied]);

  return (
    <div className="flex w-full gap-2.5 items-center mb-3">
      <div className="w-[50px] h-[50px] object-fill rounded overflow-hidden">
        <RandomAvatar name={DUMMY_WALLET_ADDRESS} variant="beam" />
      </div>
      <div className="flex flex-col w-full">
        {DUMMY_USERNAME ?? <Body className="text-white">{DUMMY_USERNAME}</Body>}
        <div className="flex gap-1.5 items-center w-full overflow-hidden">
          <p className="text-sm text-white font-light truncate">
            {DUMMY_WALLET_ADDRESS}
          </p>
          <CopyToClipboard text={DUMMY_WALLET_ADDRESS} onCopy={onCopyCodeClick}>
            <div className="cursor-pointer">
              {copied ? (
                <CopySuccessIcon size={14} color="white" />
              ) : (
                <CopyIcon size={14} color="white" />
              )}
            </div>
          </CopyToClipboard>
        </div>
      </div>
    </div>
  );
};

export default UserInfo;
