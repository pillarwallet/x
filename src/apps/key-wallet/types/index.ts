export interface Asset {
  id: number;
  name: string;
  symbol: string;
  logo: string;
  balance: number;
  decimals: number;
  price: number;
  price_change_24h: number;
  contract: string;
  chainId: number;
  chainName: string;
  usdBalance: number;
}

export interface SendFormData {
  amount: string;
  recipient: string;
}

export interface TransactionStatus {
  hash: string;
  chainId: number;
  status: 'pending' | 'success' | 'failed';
  timestamp: number;
}

