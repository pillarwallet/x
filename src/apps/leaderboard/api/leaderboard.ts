import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

// store
import { addMiddleware } from '../../../store';

// types
import { PointsResultsData } from '../../../types/api';

// utils
import { CompatibleChains, isTestnet } from '../../../utils/blockchain';

const chainIds = isTestnet
  ? [11155111]
  : CompatibleChains.map((chain) => chain.chainId);

const chainIdsQuery = chainIds.map((id) => `chainIds=${id}`).join('&');

export const leaderboardApi = createApi({
  reducerPath: 'leaderboardApi',
  baseQuery: fetchBaseQuery({
    baseUrl: isTestnet
      ? 'https://pxpoints-nubpgwxpiq-uc.a.run.app'
      : 'https://pxpoints-7eu4izffpa-uc.a.run.app',
  }),
  endpoints: (builder) => ({
    getLeaderboard: builder.query<
      PointsResultsData,
      { address?: string; from?: number; until?: number }
    >({
      query: ({ address, from, until }) =>
        `?${address ? `address=${address}&` : ''}${from ? `from=${from}&` : ''}${until ? `until=${until}&` : ''}${chainIdsQuery}&testnets=${String(isTestnet)}`,
    }),
  }),
});

/**
 * Add this to the store
 */
addMiddleware(leaderboardApi);

export const { useGetLeaderboardQuery } = leaderboardApi;
