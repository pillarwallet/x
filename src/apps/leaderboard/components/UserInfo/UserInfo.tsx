import {
  Copy as CopyIcon,
  CopySuccess as CopySuccessIcon,
} from 'iconsax-react';
import { useCallback, useEffect, useState } from 'react';
import CopyToClipboard from 'react-copy-to-clipboard';
import { TbTriangleFilled, TbTriangleInvertedFilled } from 'react-icons/tb';

// types
import { LeaderboardRankChange } from '../../../../types/api';

// utils
import { truncateAddress } from '../../../../utils/blockchain';

// components
import RandomAvatar from '../RandomAvatar/RandomAvatar';
import Body from '../Typography/Body';
import BodySmall from '../Typography/BodySmall';

type UserInfoProps = {
  rank: number;
  walletAddress: string;
  rankChange?: LeaderboardRankChange;
  username?: string;
};

const UserInfo = ({
  rank,
  username,
  walletAddress,
  rankChange,
}: UserInfoProps) => {
  const [copied, setCopied] = useState<boolean>(false);
  const [isMobile, setIsMobile] = useState<boolean>(false);

  useEffect(() => {
    const handleResize = () => {
      setIsMobile(window.innerWidth < 769);
    };

    handleResize();
    window.addEventListener('resize', handleResize);

    return () => window.removeEventListener('resize', handleResize);
  }, []);

  const onCopyCodeClick = useCallback(() => {
    if (copied) {
      setCopied(false);
      return;
    }

    setCopied(true);
  }, [copied]);

  useEffect(() => {
    const codeCopyActionTimeout = setTimeout(() => {
      setCopied(false);
    }, 500);

    return () => {
      clearTimeout(codeCopyActionTimeout);
    };
  }, [copied]);

  const rankChangeTriangle =
    rankChange === LeaderboardRankChange.DECREASED ? (
      <TbTriangleFilled
        data-testid="leaderboard-rank-up"
        color="#FF366C"
        size={9}
      />
    ) : (
      <TbTriangleInvertedFilled
        data-testid="leaderboard-rank-down"
        color="#05FFDD"
        size={9}
      />
    );

  return (
    <div
      id="leaderboard-user-data"
      data-testid="leaderboard-user-data"
      className="flex desktop:gap-3.5 tablet:gap-3.5 mobile:gap-1.5 items-center"
    >
      <div className="flex items-center desktop:gap-3.5 tablet:gap-3.5 mobile:gap-1.5">
        <div
          className={`relative group desktop:w-[35px] tablet:w-[35px] mobile:w-[30px] ${
            rankChange === LeaderboardRankChange.NO_CHANGE &&
            'desktop:mr-[23px] tablet:mr-[23px] mobile:mr-[15px]'
          }`}
        >
          <Body className="text-purple_light truncate overflow-hidden w-full mobile:text-sm">
            #{rank}
          </Body>
          <span className="absolute left-1/2 top-full mt-1 w-max max-w-xs -translate-x-1/2 scale-0 text-white text-xs group-hover:scale-100 transition">
            #{rank}
          </span>
        </div>
        {rankChange !== LeaderboardRankChange.NO_CHANGE && rankChangeTriangle}
        <div className="desktop:w-[50px] desktop:h-[50px] tablet:w-[50px] tablet:h-[50px] mobile:w-[40px] mobile:h-[40px] object-fill rounded mr-3.5 overflow-hidden">
          <RandomAvatar name={walletAddress} variant="beam" />
        </div>
      </div>
      <div className="flex flex-col w-[60%]">
        {username ?? <Body className="text-white">{username}</Body>}
        <div className="flex desktop:gap-3.5 tablet:gap-3.5 mobile:gap-1.5 items-end">
          <BodySmall className="text-purple_light">
            {isMobile
              ? `${walletAddress.slice(0, 6)}...`
              : truncateAddress(walletAddress, 18)}
          </BodySmall>
          <CopyToClipboard text={walletAddress} onCopy={onCopyCodeClick}>
            <div className="flex gap-2 items-center cursor-pointer self-end">
              {copied ? (
                <CopySuccessIcon
                  size={16}
                  color="#E2DDFF"
                  data-testid="copy-success-icon"
                />
              ) : (
                <CopyIcon size={15} color="#E2DDFF" data-testid="copy-icon" />
              )}
            </div>
          </CopyToClipboard>
        </div>
      </div>
    </div>
  );
};

export default UserInfo;
