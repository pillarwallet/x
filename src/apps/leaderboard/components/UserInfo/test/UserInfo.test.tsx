import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

// types
import { LeaderboardRankChange } from '../../../../../types/api';

// components
import UserInfo from '../UserInfo';

describe('<UserInfo />', () => {
  const mockUser = {
    rank: 1,
    walletAddress: '0x1234567890abcdef',
    rankChange: LeaderboardRankChange.INCREASED,
    username: 'User1',
  };

  it('renders correctly and matches snapshot', () => {
    const tree = renderer
      .create(
        <UserInfo
          rank={mockUser.rank}
          walletAddress={mockUser.walletAddress}
          rankChange={mockUser.rankChange}
          username={mockUser.username}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('displays the correct rank and username', () => {
    render(
      <UserInfo
        rank={mockUser.rank}
        walletAddress={mockUser.walletAddress}
        rankChange={mockUser.rankChange}
        username={mockUser.username}
      />
    );
    expect(screen.getAllByText(`#${mockUser.rank}`)).toHaveLength(2);
    expect(screen.getByText(mockUser.username)).toBeInTheDocument();
  });

  it('displays truncated wallet address', () => {
    render(
      <UserInfo
        rank={mockUser.rank}
        walletAddress={mockUser.walletAddress}
        rankChange={mockUser.rankChange}
        username={mockUser.username}
      />
    );
    expect(screen.getByText(/0x1234.*cdef/)).toBeInTheDocument();
  });

  it('shows copy success icon when copying', () => {
    render(
      <UserInfo
        rank={mockUser.rank}
        walletAddress={mockUser.walletAddress}
        rankChange={mockUser.rankChange}
        username={mockUser.username}
      />
    );
    fireEvent.click(screen.getByTestId('copy-icon'));
    expect(screen.getByTestId('copy-success-icon')).toBeInTheDocument();
  });

  it('displays the correct rank change icon if rank down', () => {
    render(
      <UserInfo
        rank={mockUser.rank}
        walletAddress={mockUser.walletAddress}
        rankChange={mockUser.rankChange}
        username={mockUser.username}
      />
    );
    expect(screen.getByTestId('leaderboard-rank-down')).toBeInTheDocument();
  });

  it('displays the correct rank change icon if rank up', () => {
    render(
      <UserInfo
        rank={mockUser.rank}
        walletAddress={mockUser.walletAddress}
        rankChange={LeaderboardRankChange.DECREASED}
        username={mockUser.username}
      />
    );
    expect(screen.getByTestId('leaderboard-rank-up')).toBeInTheDocument();
  });

  it('does not display any icon if rank change is same', () => {
    render(
      <UserInfo
        rank={mockUser.rank}
        walletAddress={mockUser.walletAddress}
        rankChange={LeaderboardRankChange.NO_CHANGE}
        username={mockUser.username}
      />
    );
    expect(screen.queryByTestId('leaderboard-rank-up')).not.toBeInTheDocument();
    expect(
      screen.queryByTestId('leaderboard-rank-down')
    ).not.toBeInTheDocument();
  });

  it('does not display username if does not exists', () => {
    render(
      <UserInfo
        rank={mockUser.rank}
        walletAddress={mockUser.walletAddress}
        rankChange={mockUser.rankChange}
      />
    );
    expect(screen.queryByText(mockUser.username)).not.toBeInTheDocument();
  });
});
