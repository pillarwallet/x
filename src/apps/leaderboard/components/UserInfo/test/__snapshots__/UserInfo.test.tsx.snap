// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<UserInfo /> > renders correctly and matches snapshot 1`] = `
<div
  className="flex desktop:gap-3.5 tablet:gap-3.5 mobile:gap-1.5 items-center"
  data-testid="leaderboard-user-data"
  id="leaderboard-user-data"
>
  <div
    className="flex items-center desktop:gap-2.5 tablet:gap-2.5 mobile:gap-1.5"
  >
    <div
      className="relative group desktop:w-[45px] tablet:w-[45px] mobile:w-[35px] flex-shrink-0"
    >
      <p
        className="text-base font-medium truncate text-white overflow-hidden mobile:text-sm w-full"
      >
        #
        1
      </p>
      <span
        className="absolute left-1/2 top-full mt-1 w-max max-w-xs -translate-x-1/2 scale-0 group-hover:scale-100 transition text-white text-xs z-10"
      >
        #
        1
      </span>
    </div>
    <div
      className="desktop:flex tablet:flex mobile:hidden w-[9px]"
    >
      <svg
        color="#FF366C"
        data-testid="leaderboard-rank-down"
        fill="none"
        height={9}
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        style={
          {
            "color": "#FF366C",
          }
        }
        viewBox="0 0 24 24"
        width={9}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
          stroke="none"
        />
        <path
          d="M20.118 3h-16.225a2.914 2.914 0 0 0 -2.503 4.371l8.116 13.549a2.917 2.917 0 0 0 4.987 .005l8.11 -13.539a2.914 2.914 0 0 0 -2.486 -4.386z"
          fill="currentColor"
          strokeWidth="0"
        />
      </svg>
    </div>
    <div
      className="flex-shrink-0 desktop:w-[50px] desktop:h-[50px] tablet:w-[50px] tablet:h-[50px] mobile:w-[40px] mobile:h-[40px] object-fill rounded overflow-hidden"
    >
      <div
        data-testid="mock-avatar"
      >
        Avatar
      </div>
    </div>
  </div>
  <div
    className="flex flex-col w-[60%]"
  >
    <p
      className="text-base font-medium text-white"
    />
    <div
      className="flex items-center desktop:w-[200px] tablet:w-[200px] mobile:w-[120px] xxs:w-[100px] justify-between gap-2"
    >
      <p
        className="text-sm font-medium font-normal text-white"
      />
      <div
        className="flex-shrink-0 w-[16px] h-[16px] flex items-center justify-center"
      >
        <div
          className="cursor-pointer"
          onClick={[Function]}
        >
          <img
            alt="copy-address"
            className="w-[13px] h-[15px]"
            src="/src/apps/pillarx-app/images/token-market-data-copy.png"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
